File: simlf/src/basic/sup/univ/sup_univ_cn.py
import logging
import numpy as np
from sim import Module

class SupUnivCn(Module):
    def run_impl(self):
        #hash_int_set = yang::unordered_set<int>;
        univ = self.config["par_univ"]
        b_univ_custom = self.write_array(univ, dtype=bool)
        b_sup_passed = self.write_array(univ + "_passed", dtype=bool)
        univ_size_ = int(self.config["par_univ_size"])
        lookback_ = int(self.config["par_univ_lookback"])
        sticky_ = int(self.config["par_univ_sticky_days"])
        vol_window_ = int(self.config["par_univ_vol_window"])
        pass_rate_ = float(self.config["par_univ_pass_rate"])
        min_cap_ = float(self.config["par_univ_min_cap"])
        min_dvol_ = float(self.config["par_univ_min_dvol"])
        min_prc_ = float(self.config["par_univ_min_prc"])
        max_prc_ = float(self.config["par_univ_max_prc"])
        vol_mode_ = "mean"

        b_univ_parent = self.read_array(self.config.get("par_univ_parent", "base/univ_all"))
        candidates = {}
        b_close = self.read_array("base/close")
        b_vol = self.read_array("base/vol")
        b_cap = self.read_array("base/cap")
        b_st = self.read_array("base/st")
        get_dollar_vol = lambda di, ii: b_close[di, ii] * b_vol[di, ii]

        for di in range(self.start_di - lookback_ if self.start_di > lookback_ else 1, self.end_di):
            sorted_vol = {}
            for ii in range(self.univ_size):
                fx = 1.
                if b_univ_parent[di, ii] and b_st[di, ii] == 0 and b_cap[di - 1, ii] / fx >= min_cap_ and b_close[di - 1, ii] / fx >= min_prc_ and b_close[di - 1, ii] / fx <= max_prc_:
                    dvols = []
                    for w in range(vol_window_):
                        if w > di - 1:
                            break
                        dvol = get_dollar_vol(di - 1 - w, ii)
                        if dvol >= 0 and np.isfinite(dvol):
                            dvols.append(dvol)
                    if dvols:
                        est_dvol = self.CalcEstimatedVol(dvols, vol_mode_)
                        if est_dvol >= min_dvol_:
                            sorted_vol[est_dvol] = ii

#       // select the top (univ_size) stocks as candidates
#       // traversing sorted_vol backwards gives us the avg volumes in
#       // decreasing order
            candidates[di] = set()
            result = candidates[di]
            for rcurr in sorted(sorted_vol, reverse=True):
                if len(result) < univ_size_:
                    result.add(sorted_vol[rcurr])
            if not result:
                logging.error(f"[{self.name}] no stocks matched universe volume criteria on {self.dates[di]}")
            else:
                logging.info(f"[{self.name}] Selected {len(result)} candidates on {self.dates[di]} in the first pass")

#     // 2nd pass
        pick_cnt = [0] * self.univ_size
        last_good = [-1] * self.univ_size
        for ii in range(self.univ_size):
            di = self.start_di - 1
            if di >= 0:
                lb1 = lookback_ if di > lookback_ else di
                lb2 = sticky_ if di > sticky_ else di
                for dl in range(lb1):
                    if ii in candidates[di - dl]:
                        pick_cnt[ii] += 1
                for dl in range(lb2):
                    if b_sup_passed(di - dl, ii):
                        last_good[ii] = di - dl
                        break

        for di in range(self.start_di, self.end_di):
            if di < 1:  # this selection has to always be delay 1
                continue
            lb1 = lookback_ if di > lookback_ else di
            lb2 = sticky_ if di > sticky_ else di
            dl = 0
#       // error check: make sure all necessary lookback dates were covered
#       // in 1st pass
            for dl in range(lb1):
                if di - dl not in candidates:
                    logging.warning(f"[{self.name}] universe: candidate universe was not computed for {di - 1 - dl}")

            activated = added = removed = 0
            for ii in range(self.univ_size):
#         // master universe gets priority
                good = b_univ_parent[di, ii]
                if di > lookback_ and pick_cnt[ii] > 0:
                    pick_cnt[ii] -= ii in candidates[di - lookback_]
                if good:
#           // check the last 'lookback' days
#           // candidates was computed with delay 1 info, so we can
#           // start with today
                    if ii in candidates[di]:
                        pick_cnt[ii] += 1
                    assert pick_cnt[ii] <= lb1
                    if lb1 > 0 and pick_cnt[ii] / lb1 * 100 < pass_rate_:
                        good = False
#           // record whether it passed_ today
                    b_sup_passed[di, ii] = good
                    if good:
                        last_good[ii] = di
                    else:
#             // must be bad for 'sticky' days in order to deactivate
                        good = last_good[ii] >= di - lb2
                if good:
                    activated += 1
                    b_univ_custom[di, ii] = True
                    if di == 0 or not b_univ_custom[di - 1, ii]:
                        added += 1
                else:
                    b_univ_custom[di, ii] = False
                    if di == 0 or b_univ_custom[di - 1, ii]:
                        removed += 1

            logging.info(f"[{self.name}] Activated {activated} instruments on {self.dates[di]} (+{added} -{removed})")

    def CalcEstimatedVol(self, vols, vol_mode):
        est_vol = 0.0
        def find_median():
            n = len(vols) // 2
            vols.sort()
            if len(vols) % 2 == 0:
                return (vols[n] + max(vols[:n])) / 2
            else:
                return vols[n]
        def find_mean():
            sum_ = 0
            for v in vols:
                sum_ += v
            return sum_ / len(vols)

        if vol_mode == "median":
            est_vol = find_median()
        elif vol_mode == "min_median_mean":
            est_vol = min(find_median(), find_mean())
        else:
            est_vol = find_mean()

        return est_vol

File: simlf/src/basic/sup/univ/sup_univ_cn.cpp
#include "yang/util/unordered_set.h"
#include "yao/base_daily.h"
using namespace yao;

namespace {
using namespace std;
class SupUnivCn : public yang::Module {
 public:
  void RunImpl() final {
    using hash_int_set = yang::unordered_set<int>;

    auto univ = config<std::string>("par_univ");
    auto b_univ_custom = WriteArray<bool>(univ);
    auto b_sup_passed = WriteArray<bool>(univ + "_passed");
    int univ_size_ = config<int>("par_univ_size");
    int lookback_ = config<int>("par_univ_lookback");
    int sticky_ = config<int>("par_univ_sticky_days");
    int vol_window_ = config<int>("par_univ_vol_window");
    float pass_rate_ = config<float>("par_univ_pass_rate");
    float min_cap_ = config<float>("par_univ_min_cap");
    float min_dvol_ = config<float>("par_univ_min_dvol");
    float min_prc_ = config<float>("par_univ_min_prc");
    float max_prc_ = config<float>("par_univ_max_prc");
    string vol_mode_ = "mean";

    auto &b_univ_parent = *ReadArray<bool>(config<string>("par_univ_parent", "base/univ_all"));

    std::map<int, hash_int_set> candidates;  // map of date => stocks in candidate universe
    auto &b_close = *ReadArray<float>("base/close");
    auto &b_vol = *ReadArray<float>("base/vol");
    auto &b_cap = *ReadArray<float>("base/cap");
    auto &b_st = *ReadArray<int>("base/st");

    auto get_dollar_vol = [&](int di, int ii) {
      // int currency = (*currency_)(di, ii);
      // if (IsNull(currency)) return Null<float>();

      // float fxrate = (*fxrate_)(di, currency);
      float fxrate = 1;
      float price = b_close(di, ii);
      float volume = b_vol(di, ii);

      return price * volume / fxrate;
    };

    for (int di = (start_di() > lookback_ ? start_di() - lookback_ : 1); di < end_di(); ++di) {
      std::multimap<float, int> sorted_vol;  // sorted map of avg_vol => instrument
      for (int ii = 0; ii < univ_size(); ++ii) {
        // filter by exchange

        float fx = 1;

        if (b_univ_parent(di, ii) && b_st(di, ii) == 0 && my_valid(b_cap(di - 1, ii)) &&
            (b_cap(di - 1, ii) / fx >= min_cap_) && (b_close(di - 1, ii) / fx >= min_prc_) &&
            (b_close(di - 1, ii) / fx <= max_prc_)) {
          // compute trailing average notional volume for current
          // instrument

          vector<float> dvols;
          for (int w = 0; w < vol_window_; ++w) {
            if (w > di - 1) break;
            float dvol = get_dollar_vol(di - 1 - w, ii);
            if (dvol >= 0 && my_valid(dvol)) {
              dvols.push_back(dvol);
            }
          }

          if (dvols.size() > 0) {
            float est_dvol = CalcEstimatedVol(dvols, vol_mode_);

            if (est_dvol >= min_dvol_) sorted_vol.insert(std::make_pair(est_dvol, ii));
          }
        }
      }

      // select the top (univ_size) stocks as candidates
      // traversing sorted_vol backwards gives us the avg volumes in
      // decreasing order
      candidates.insert(std::make_pair(di, hash_int_set(univ_size_)));
      hash_int_set &result = candidates[di];
      for (std::multimap<float, int>::reverse_iterator rcurr = sorted_vol.rbegin();
           rcurr != sorted_vol.rend() && static_cast<int>(result.size()) < univ_size_; ++rcurr) {
        result.insert(rcurr->second);
      }
      if (result.empty()) {
        LOG_ERROR("[{}] no stocks matched universe volume criteria on {}", name(), dates()[di]);
      } else {
        LOG_INFO("[{}] Selected {} candidates on {} in the first pass", name(), result.size(),
                 dates()[di]);
      }
    }

    // 2nd pass
    std::vector<int> pick_cnt(univ_size());
    std::vector<int> last_good(univ_size(), -1);
    for (int ii = 0; ii < univ_size(); ++ii) {
      const int di = start_di() - 1;
      if (di < 0) continue;
      const int lb1 = (di > lookback_) ? lookback_ : di;
      const int lb2 = (di > sticky_) ? sticky_ : di;
      for (int dl = 0; dl < lb1; ++dl) {
        pick_cnt[ii] += candidates[di - dl].find(ii) != candidates[di - dl].end();
      }
      for (int dl = 0; dl < lb2; ++dl) {
        if (b_sup_passed(di - dl, ii)) {
          last_good[ii] = di - dl;
          break;
        }
      }
    }
    for (int di = start_di(); di < end_di(); ++di) {
      if (di < 1) continue;  // this selection has to always be delay 1

      const int lb1 = (di > lookback_) ? lookback_ : di;
      const int lb2 = (di > sticky_) ? sticky_ : di;

      // error check: make sure all necessary lookback dates were covered
      // in 1st pass
      int dl;
      for (dl = 0; dl < lb1; ++dl) {
        if (candidates.find(di - dl) == candidates.end()) {
          LOG_WARN("[{}] universe: candidate universe was not computed for {}", name(),
                   di - 1 - dl);
        }
      }

      int activated = 0, added = 0, removed = 0;
      for (int ii = 0; ii < univ_size(); ++ii) {
        // master universe gets priority

        bool good = b_univ_parent(di, ii);
        if (di > lookback_ && pick_cnt[ii] > 0) {
          pick_cnt[ii] -= candidates[di - lookback_].find(ii) != candidates[di - lookback_].end();
        }

        if (good) {
          // check the last 'lookback' days
          // candidates was computed with delay 1 info, so we can
          // start with today
          if (candidates[di].find(ii) != candidates[di].end()) ++pick_cnt[ii];
          ENSURE2(pick_cnt[ii] <= lb1);

          if (lb1 > 0 && static_cast<double>(pick_cnt[ii]) / lb1 * 100. < pass_rate_) good = false;

          // record whether it passed_ today
          b_sup_passed(di, ii) = good;

          if (good) {
            last_good[ii] = di;
          } else {
            // must be bad for 'sticky' days in order to deactivate
            good = last_good[ii] >= di - lb2;
          }
        }

        if (good) {
          ++activated;
          b_univ_custom(di, ii) = true;
          if (0 == di || !b_univ_custom(di - 1, ii)) ++added;
        } else {
          b_univ_custom(di, ii) = false;
          if (0 == di || b_univ_custom(di - 1, ii)) ++removed;
        }
      }

      LOG_INFO("[{}] Activated {} instruments on {} (+{} -{})", name(), activated,
               env().dates()[di], added, removed);
    }
  }

  float CalcEstimatedVol(vector<float> &vols, string vol_mode_) {
    float est_vol = 0.0;
    auto find_median = [&]() {
      int n = vols.size() / 2;
      std::nth_element(vols.begin(), vols.begin() + n, vols.end());
      if (vols.size() % 2 == 0) {
        return (vols[n] + *std::max_element(vols.begin(), vols.begin() + n)) / 2;
      } else {
        return vols[n];
      }
    };
    auto find_mean = [&]() {
      float sum = 0;
      for (auto &v : vols) {
        sum += v;
      }
      return sum / vols.size();
    };

    if (vol_mode_ == "median") {
      est_vol = find_median();
    } else if (vol_mode_ == "min_median_mean") {
      est_vol = std::min(find_median(), find_mean());
    } else {
      est_vol = find_mean();
    }

    return est_vol;
  }
};

REGISTER_MODULE(SupUnivCn);
}  // namespace
